[FLINK-15853][hive][table-planner-blink] Use the new type inference f…

lirui-apache opened the pull request:
…or hive udf

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Migrate Hive UDF to new type inference


## Brief change log

  - Return `HiveSimpleUDF/HiveGenericUDF` instead of `ScalarFunctionDefinition` in `HiveFunctionDefinitionFactory`
  - Implement a `TypeStrategy` for `HiveScalarFunction` and use it to create the `TypeInference`. This `TypeStrategy` takes care of setting argument types & constants for the Hive function.


## Verifying this change

Existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

twalthr commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -82,6 +84,13 @@ public void open(FunctionContext context) {
+	@Override
+	public TypeInference getTypeInference(DataTypeFactory typeFactory) {
+		TypeInference.Builder builder = TypeInference.newBuilder();
+		builder.outputTypeStrategy(new ResultTypeStrategy(this));
```
I guess we will also need a matching input type strategy such that the validation can work properly.

lirui-apache commented:
@twalthr Thanks for the review. Currently the input validation happens when inferring the result type of the UDF, i.e. an exception is thrown if we cannot find a proper Hive function for the input arguments.

I noticed both `InputTypeStrategy::inferInputTypes` and `TypeStrategy::inferType` takes a `CallContext` argument. I wonder whether/how these two `CallContext` are related. For example, if I implement `InputTypeStrategy::inferInputTypes` and validate the input, then when I implement `TypeStrategy::inferType` do I need to validate the input again? Or is it guaranteed that the input has already been validated at this point?

lirui-apache commented:
@twalthr I have rebased and implemented an input strategy. The input strategy will attempt to get a proper Hive function according to the input arguments. The input is considered valid if such a Hive function can be found. And this Hive function also gives us the result type of the function call. So the output strategy can simply return this result type -- here I'm assuming `InputTypeStrategy::inferInputTypes` and `TypeStrategy::inferType` see the same `CallContext`.  Please let me know if this is the right way to go.

> remove all Hive related code in Blink planner

I have removed `HiveScalarSqlFunction`. We'll get rid of all the Hive code once we update Hive UDAF and UDTF as well.

JingsongLi commented:
Hi @lirui-apache , can we let `setArgumentTypesAndConstants` and `getHiveResultType` in `HiveFunction` be default function, default throw exception.
And we can have a clean `HiveScalarFunction`, remove `getResultType` and other legacy methods.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +142,62 @@ public Object eval(Object... args) {
+	/**
+	 * Validate input argument types and decide result type.
+	 */
+	protected abstract DataType validateInputTypes(DataType[] argTypes) throws UDFArgumentException;
```
Looks like it is `inferOutputType`

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +142,62 @@ public Object eval(Object... args) {
+		}
+	}
+
+	private class HiveUDFInputStrategy implements InputTypeStrategy {
```
I can not get it what this is for.

lirui-apache replied:
I think this is needed for input type validation. For example, according to the API contract, a `ValidationException` can be thrown if the input is invalid, but this should only happen during input inference rather than output inference.

JingsongLi replied:
Got it.

lirui-apache commented:
@JingsongLi Thanks for the review. I have updated the PR including:
1. `HiveScalarFunction` no longer implements `HiveFunction`. I believe we can get rid of this interface once we migrate UDAF & UDTF to new type inference.
2. Removed some legacy methods.
3. Some refactors

JingsongLi reviewed (commented):
Looks good to me, @twalthr Do you have other comments?

JingsongLi reviewed (approved):
Looks good to me.

twalthr commented:
Sorry, for the late response. I was on vacation. I will leave some final feedback that we can address in FLINK-15854. Are you also taking care of this issue @lirui-apache?

twalthr commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +114,66 @@ public Object eval(Object... args) {
 	 */
 	protected abstract Object evalInternal(Object[] args);
+
+	private void setArguments(CallContext callContext) {
```
Move this to a dedicated class because it will be shared for other functions as well?

JingsongLi replied:
We can rename `HiveFunction` to `HiveLegacyFunction`, and create a new abstract base class `HiveFunction`.

twalthr commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +114,66 @@ public Object eval(Object... args) {
+		DataType[] inputTypes = callContext.getArgumentDataTypes().toArray(new DataType[0]);
+		Object[] constantArgs = new Object[inputTypes.length];
+		for (int i = 0; i < constantArgs.length; i++) {
+			if (callContext.isArgumentLiteral(i)) {
```
We should also check for `isNullLiteral`?

JingsongLi replied:
Looks like `getArgumentValue` will return empty for null literal. But we should add test for null literal.

twalthr commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +114,66 @@ public Object eval(Object... args) {
+				inferReturnType();
+			} catch (UDFArgumentException e) {
+				if (throwOnFailure) {
+					throw new ValidationException(
```
the `CallContext` provides a method for throwing validation errors.

twalthr commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java:
```diff
@@ -104,4 +114,66 @@ public Object eval(Object... args) {
+
+		@Override
+		public List<Signature> getExpectedSignatures(FunctionDefinition definition) {
+			return Collections.singletonList(Signature.of(Signature.Argument.of("*")));
```
I think it would be helpful to print the expected signature in case the user passes wrong arguments?

lirui-apache replied:
I'm not sure whether it's possible to get all the valid signatures for a hive function. Some hive function supports implicit type conversion, e.g. a function that operates on string values may take numeric input as well. I'll double check.

twalthr replied:
Actually, this is why this method is very generic and allows you to pass arbitrary strings like `NUMERIC` or `ANY`. I don't know how Hive implements UDFs but as a user I would find it useful to get exceptions like
```
Invalid input arguments. Expected signatures are:\n
f(STRING NOT NULL)\nf(INT NOT NULL)\nf(BIGINT NOT NULL)
```

Btw you can also use `InputTypeStrategiesTestBase` for testing. See `InputTypeStrategiesTest` for an example.

twalthr commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java:
```diff
@@ -291,4 +297,58 @@ public String evaluate(String content) {
+	/**
+	 * A CallContext implementation for Hive UDF tests.
+	 */
+	public static class HiveUDFCallContext implements CallContext {
```
Use `org.apache.flink.table.types.inference.utils.CallContextMock`

JingsongLi commented:
> Sorry, for the late response. I was on vacation. I will leave some final feedback that we can address in [FLINK-15854](https://issues.apache.org/jira/browse/FLINK-15854). Are you also taking care of this issue @lirui-apache?

Hi @twalthr , thanks for review again. @lirui-apache has two days off, but he will continue to push forward the inference issues when he comes back.

lirui-apache commented:
@twalthr Thanks for the comments. I'll address them when I work on FLINK-15854
